diff --git a/src/wrapped/wrappedlibdl.c b/src/wrapped/wrappedlibdl.c
index 0d69d80..bf59f9e 100644
--- a/src/wrapped/wrappedlibdl.c
+++ b/src/wrapped/wrappedlibdl.c
@@ -245,8 +245,8 @@ static void* my_dlopen_internal(x64emu_t* emu, void *filename, int flag)
             return NULL;
         }
         setupTraceMapLib(tmp->libs[0]->maplib);
+        lib = tmp->libs[0];
         free_neededlib(tmp);
-        lib = GetLibInternal(rfilename);
         refreshTLSData(emu);
         RunDeferredElfInit(emu);
         my_context->deferredInit = old_deferredInit;
